### ISSUE TITLE
Allow Greenmail to be configured as mail sink

### ISSUE THREAD

**Issue by msaladin** [greenmail-mail-test/greenmail#165]
In our company we have a lot of projects which currently use James as a test-mail-server for their test- and integration environments. Those projects want to:
- make sure that mails are not forwarded to the real recipient
- but they still want to be able to browse the emails which were sent by the SUT

To do this, they have configured an email client which connects to James and uses IMAP to retrieve messages from one IMAP account, and James is configured as mail-sink to deliver all the mails to this one IMAP account. We would like to use Greenmail instead, and it looks like this is not really implemented currently:

I suggest to enhance Greenmail with:
- Property which can be set to set that Greemail should be acting as mailsink (Boolean property: greenmail.act-as-mailsink, default: false)
- Property with username for the mailsink (String property: greenmail.mailsink-username, default username: mailsink)
- Incoming mails (by SMTP) would then be routed to only one user (one IMAP mailbox.INBOX)

The changes to the code would be small, mostly to class SmtpManager.Incoming and to configuration-property handling. I could contribute this if this makes sense for the maintainers.

Any feedback is appreciated.
Cheers Michael

**Comment by camann9** [comment]
@marcelmay What do you think?
I think there is little to be said against this. We'd just have to think about how we configure this

**Comment by buildscientist** [comment]
It seems like we already have this feature in place - meaning do you really the need the email to be "forwarded" on to another recipient or are you just looking for the mail-sink functionality? 

We already have a mailsink in the form of of our InMemoryStore. So you could start up a GreenMail SMTP service - have the email relayed and then access it via the InMemoryStore. 

It's not the most convenient way of accessing your mail-sink but it works.

**Comment by msaladin** [comment]
@buildscientist: In our use-case from above, the access to the mailsink is done using an IMAP client, e.g. by a human person using a mail client. For example:

- Maybe by a manual tester wants to check whether the manual tests he has done on the SUT generate the correct emails with correct substitutions.

- Or the real customer wants to click around in the integration environment and then look at all the emails that were sent to all different stakeholders (but not really sent, because they were sent to Greenmail and stored inside a single IMAP Inbox).

**Comment by buildscientist** [comment]
@msaladin - makes sense but it sounds like you're now using GreenMail as a workflow engine. It wasn't really designed as such. 

I think the idea of having a GreenMail config that allows users to forward email to other users is a valid feature though.

**Comment by camann9** [comment]
I think it's still a reasonable thing to do, even if it wasn't designed as
such.
So I'd be good with the forwarding feature as long as it has sufficient
test coverage.

**Comment by marcelmay** [comment]
@msaladin, I still have some questions.

* Would you only have one property of the receiving mail account for configuration?
e.g. property _greenmail.mailsink.user=foo:pwd@bar.com_ _ similiar to _greenmail.users_ configuration?

* Would this user receive a copy, or would you actually move the messages?
   A copy is less disruptive in GreenMails behaviour. So you actually would have some kind of automatic BCC to a configurable GreenMail mail account. Would this work?

**Comment by msaladin** [comment]
Good idea, I did a first implementation, see here:
https://github.com/greenmail-mail-test/greenmail/pull/167

There were changes to more files than I had imagined in the first view.

Mainly because now some of the "managers" needed the configuration, and for this, it was necessary to make a clear distinction between an instance of GreenMail which is not yet started and which is started. The idea is that when the GreenMail instance is not yet started, the properties can be changed from outside, like this:

`GreenMail greenMail = new GreenMail(ServerSetupTest.ALL);`
`greenMail.withConfiguration(new PropertiesBasedGreenMailConfigurationBuilder().build(props));`
`greenMail.start();`

I added the initialization of the managers to the start() command, and made sure that no client calls the setUser() method before starting GreenMail. You see that only at the time that start() is invoked it it clear which startup Properties to take.

The real business logic is small enough and is in class SmtpManager.Incoming.
